core/common/iterator/src/main/java/org/eclipse/rdf4j/common/iteration/CloseableIteration.java

```diff
@@ -23,9 +27,32 @@
  *   // process the exception that can be thrown while processing.
  * }
  * </pre>
+ *
+ * @implNote In the future this interface will stop extending {@link Iteration} and instead declare the same interface
+ *           methods directly. The interface will also stop requiring implementations to automatically close when
+ *           exhausted, instead making this an optional feature and requiring the user to always call close.
```

Comment:
I think it's worth also marking this interface as `@Deprecated` (though not necessarily for removal), and start having a think about replacing the whole concept of Iterations with something else, for example something like (just spitballing here):

```java
public interface ResultIterator<T> extends Iterator<T>, AutoCloseable 
```

Not saying we _do_ want to go that way, but making this deprecated now gives us options in 5.0.

Reply:
Sounds good.